<div align=left><img width=30% height=30% src="https://github.com/xjtu-omics/SVision/blob/master/supports/svision-logo.png"/></div>


SVision is a deep learning-based structural variants caller that takes aligned reads or contigs as input. 
Especially, SVision implements a targeted multi-objects recognition framework, detecting and characterizing both simple and complex structural variants from three-channel similarity images.

<div align=left><img width=80% height=80% src="https://github.com/xjtu-omics/SVision/blob/master/supports/workflow.png"/></div> 


## License and citation

SVision is free for non-commercial use by academic, government, and non-profit/not-for-profit institutions. A commercial version of the software is available and licensed through Xiâ€™an Jiaotong University. 
For more information, please contact with Jiadong Lin (jiadong66@stu.xjtu.edu.cn) or Kai Ye (kaiye@xjtu.edu.cn).

Please cite our paper "Lin, J., Wang, S., Audano, P.A. et al. SVision: a deep learning approach to resolve complex structural variants. Nat Methods (2022)." [PDF](https://www.nature.com/articles/s41592-022-01609-w.pdf)

## Installation


### Operation systems

* MacOS, Big Sur (V11.6)
* Ubuntu (V20.04, including Windows Subsystem for Linux)
* CentOS Linux (V7.6.1810)

### From source

```
## Get latest source code
git clone https://github.com/xjtu-omics/SVision.git
cd SVision

## Create conda environment and install SVision 
conda env create -f environment.yml
python setup.py install
```

### Docker

```
docker pull jiadongxjtu/svision:latest
```


## Usage

Please check the [wiki](https://github.com/xjtu-omics/SVision/wiki) page for more usage and output file format.

We provided support scripts used in this study to filter SVision calls at [SVisionUtil](https://github.com/jiadong324/SVisionUtils), please follow instructions to filter your own calls.

### General usage

#### Required Input/Ouput parameters

```
-o OUT_PATH           Absolute path to output
-b BAM_PATH           Absolute path to bam file
-m MODEL_PATH         Absolute path to CNN predict model
-g GENOME             Absolute path to your reference genome (.fai required in the directory)
-n SAMPLE             Name of the BAM sample name
```

```-g``` path to the reference genome, the index file should under the same directory. Please include all chromosomes you want to detect in the reference file. SVision only call SVs from chromosomes specified in the reference.

```-m``` path to the pre-trained deep learning model. **NOTE:** Please use *-m svision-cnn-model.ckpt* while running your data.


**NOTE:** If your input contains the alignment of assemblies, please activate the contig mode with ```--contig```. 


### Run demo data

The demo data is ./supports/HG00733.svision.demo.bam, which is extracted from whole genome sequencing HiFi 
data of [HG00733](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/working/20190925_PUR_PacBio_HiFi/) used in this study. 
We provided HG00733 whole genome calls used in this study, which is available at [SVisionUtil](https://github.com/jiadong324/SVisionUtils).
The HiFi data of HG00733 is generated by HGSVC in a recent study published at [Science](https://www.science.org/doi/10.1126/science.abf7117?url_ver=Z39.88-2003&rfr_id=ori:rid:crossref.org&rfr_dat=cr_pub%20%200pubmed).

#### Prepare required inputs

1. Download reference genome [GRCh38](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa)

2. Download pretrained [CNN model](https://drive.google.com/drive/folders/1j74IN6kPKEx9hy3aENx3zHYPUnyYWGvj?usp=sharing). 
   There are three files and please put all files under a directory (e.g., /home/user/svision_model/).



#### Run SVision

Before running, please create a directory for SVision output (e.g., /home/user/svision_out).

```
SVision -o ./home/user/svision_out -b ./supports/HG00733.svision.demo.bam -m /home/user/svision_model/svision-cnn-model.ckpt -g /path/to/reference.fa -n HG00733 -s 5 --graph --qname
```

```
docker run -v /local/path:/container/path jiadongxjtu/svision:latest SVision -o /container/path/svision_out -b /container/path/HG00733.svision.demo.bam -m /container/path/svision_model/svision-cnn-model.ckpt -g /container/path/reference.fa -n HG00733 -s 5 --graph --qname
```
**REMINDER:** For Docker run, please put your BAM file, reference file and the pre-trained model under the /local/path.
