<div align=left><img width=30% height=30% src="https://github.com/xjtu-omics/SVision/blob/master/supports/svision-logo.png"/></div>


SVision is a deep learning-based structural variants caller that takes aligned reads or contigs as input. 
Especially, SVision implements a targeted multi-objects recognition framework, detecting and characterizing both simple and complex structural variants from three-channel similarity images.

<div align=left><img width=80% height=80% src="https://github.com/xjtu-omics/SVision/blob/master/supports/workflow.png"/></div> 


## License

SVision is free for non-commercial use by academic, government, and non-profit/not-for-profit institutions. A commercial version of the software is available and licensed through Xiâ€™an Jiaotong University. 
For more information, please contact with Jiadong Lin (jiadong324@stu.xjtu.edu.cn) or Kai Ye (kaiye@xjtu.edu.cn).

## Installation


### Operation systems

* MacOS, Big Sur (V11.6)
* Ubuntu (V20.04, including Windows Subsystem for Linux)
* CentOS Linux (V7.6.1810)

### Install from source

```
## Get the source code
git clone https://github.com/xjtu-omics/SVision.git
cd SVision
## Create a conda environment for SVision
conda env create -f ./environment.yml 
## Install from source
conda activate svisionenv
python setup.py install
```

### Use docker

```
docker pull jiadongxjtu/svision:1.3.6
```


## Usage

We provided support scripts used in this study to filter SVision calls at [SVisionUtil](https://github.com/jiadong324/SVisionUtils), please follow instructions to filter your own calls.

### Run with docker

```
docker run jiadongxjtu/svision:1.3.6 SVision -h
```

### Run demo data

The demo data is ./supports/HG00733.svision.demo.bam, which is extracted from whole genome sequencing HiFi 
data of [HG00733](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/working/20190925_PUR_PacBio_HiFi/) used in this study. 
We provided HG00733 whole genome calls used in this study, which is available at [SVisionUtil](https://github.com/jiadong324/SVisionUtils).
The HiFi data of HG00733 is generated by HGSVC in a recent study published at [Science](https://www.science.org/doi/10.1126/science.abf7117?url_ver=Z39.88-2003&rfr_id=ori:rid:crossref.org&rfr_dat=cr_pub%20%200pubmed).

#### Run with graph option

1. Download reference genome [GRCh38](https://drive.google.com/file/d/10q3hEroEpGw_wqv3qvg9omF4w2kPqDm9/view?usp=sharing)

2. Download pretrained [CNN model](https://drive.google.com/drive/folders/1j74IN6kPKEx9hy3aENx3zHYPUnyYWGvj?usp=sharing). 
   There are three files and please put all files under a directory (e.g., /home/user/svision_model/).

3. Run SVision

Before running, please create a directory for SVision output (e.g., /home/user/svision_out).

```
SVision -o ./home/user/svision_out -b ./supports/HG00733.svision.demo.bam -m /home/user/svision_model/svision-cnn-model.ckpt -g /path/to/reference.fa -n HG00733 -s 5 --graph --qname
```

**NOTE:** Please DO NOT add any file extension to *svision-cnn-model.ckpt* while running SVision.

4. Output files

All outputs are under directory /home/user/svision_out, including

* ``` *.graph.vcf ``` The standard VCF output with CSV graph info columns.

* ```*.graph_exactly_match.txt ``` Identified isomorphic CSV graphs.

* ```*.graph_symmetry_match.txt``` Identified isomorphic graphs from symmetric graph topologies.

* ```graphs``` The directory for CSV graph in GFA format.

In the VCF file, you will find the insertion-inversion-insertion event at chr9:74,283,222-74,283,473, and the diagram of this event is shown in Extended Data Fig.9 in our paper.

Please use GFA_FILE_ID to find a CSV's corresponding GFA file under ./graphs

### Run your own data

```
SVision [parameters] -o <output path> -b <input bam path> -g <reference> -m <model path> -n <sample name>
```

#### Check all parameters with

```
SVision -h
```

Required Input/Ouput parameters

```
-o OUT_PATH           Absolute path to output
-b BAM_PATH           Absolute path to bam file
-m MODEL_PATH         Absolute path to CNN predict model
-g GENOME             Absolute path to your reference genome (.fai required in the directory)
-n SAMPLE             Name of the BAM sample name
```

```-g``` path to the reference genome, the index file should under the same directory.

```-m``` path to the pre-trained deep learning model. **NOTE:** Please use *-m svision-cnn-model.ckpt* while running your data.

Please check the [wiki](https://github.com/xjtu-omics/SVision/wiki) page for more usage and performance evaluation.

## Contact
If you have any questions, please feel free to contact: jiadong66@stu.xjtu.edu.cn, songbowang125@163.com
